Added round_corners parameter that allow to draw rectangles with round corners

gonzalobarbaran opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Some guidelines are provided like this, in HTML comments, to expedite the code review before merging your contribution.

First, please make sure that you have read the documentation page on fpdf2 development:
https://pyfpdf.github.io/fpdf2/Development.html

If you're new to contributing to open source projects,
you might find this free video course helpful: http://kcd.im/pull-request
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

#427 I added a `round_corners` parameter to FPDF.rect <!-- This will automatically close issue #0 once the PR is merged: https://help.github.com/en/articles/closing-issues-using-keywords -->
This parameter allows to draw a rectangle with one or more round corners. Here is an example:
 
![image](https://user-images.githubusercontent.com/59395855/169507668-b17f5e3c-3bde-42c2-a485-f4a0c0523edb.png)

These rectangles was made by the following code:

```
y = 10
pdf.rect(60, y, 33, 28, round_corners=True, style="D")

pdf.set_fill_color(0, 255, 0)
pdf.rect(100, y, 50, 10, round_corners=("BOTTOM_RIGHT"), style="DF")

pdf.set_fill_color(255, 255, 0)
pdf.rect(160, y, 10, 10, round_corners=("TOP_LEFT", "BOTTOM_LEFT"), style="F") 
```

<!-- Have you done all of these things?  -->
**Checklist**:

<!-- To check an item, place an "x" in the box like so: "- [x] Item description"
     Add "N/A" to the end of each line that's irrelevant to your changes -->

- [x] The GitHub pipeline is OK (green), <!-- The maintainers will trigger it if this is your 1st contribution -->
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged <!-- In your opinion, can this be merged as soon as it's reviewed? Else, this can be turned into a Draft PR -->

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder


- [ ] A mention of the change is present in `CHANGELOG.md`

- [x]  The PR description or comment contains a picture of a cute animal (not mandatory but encouraged 😉)

A photo of my cat doing cat stuff 😋
![gatito](https://user-images.githubusercontent.com/59395855/169510224-86d23e45-60b3-4a6a-aed2-0887e3b4b9d7.PNG)

<!-- Feel free to add additional comments, and to ask questions if some of those guidelines are unclear to you! -->

<!--
Once a PR is merged, maintainers will add your name to the contributors table in README.md.
If they forget, or you do not wish this to happen, please mention it.
-->

gonzalobarbaran commented:
Sorry if there's something that i didn't do well, it's my first time contributing on a open source project :D

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -1266,17 +1267,116 @@ def rect(self, x, y, w, h, style=None):
+        if round_corners is not False:
+            self._draw_rounded_rect(x, y, w, h, style, round_corners)
+        else:
+            style = RenderStyle.coerce(style)
```
This should be called before the `if` to ensure that `style` has a valid value that can be converted to a `RenderStyle`

gonzalobarbaran replied:
Fixed

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -1266,17 +1267,116 @@ def rect(self, x, y, w, h, style=None):
+        coor_x = [x, x + w, x, x + w]
+        coor_y = [y, y, y + h, y + h]
+
+        if round_corners is True:
```
The logic in this `if` is good.
Now if `round_corners` is neither `True` nor `False`, a check should be made to detect invalid values.
Something like `round_corners = tuple(Corner.coerce(rc) for rc in round_corners)`

gonzalobarbaran replied:
I fixed that but i had a problem, when I call the method with just one corner the `for` split the `string` . To fix that i made the next code 
```
if len(round_corners) > 4: #here is the hardcode to detect when round_corners have just one corner
    round_corners = (round_corners,)
round_corners = tuple(Corner.coerce(rc) for rc in round_corners)
```
but I think is a bit hardcoded, do you think some other solution to solve this?

Lucas-C replied:
> do you think some other solution to solve this?

Yes, `round_corners` should simply always be an iterable (a list or a tuple).
Single values should be passed this way: `round_corners=("BOTTOM_RIGHT",)`

gmischler replied:
> I fixed that but i had a problem, when I call the method with just one corner the `for` split the `string` .

The pythonic fix would be to check for the type of the argument:

```python
    if isinstance(round_corners, (str, Corner)):  # single item
        round_corners = (Corner.coerce(round_corners),)
    else:
        round_corners = [Corner.coerce(rc) for rc in round_corners]
```

Btw.: Converting the last case into a tuple is just an unnecessary extra processing step with no real benefits.

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -1266,17 +1267,116 @@ def rect(self, x, y, w, h, style=None):
+                    original_color[0], original_color[1], original_color[2]
+                )
+
+        if style != "F":
```
This means that style is either `DRAW` or `DRAW_FILL`.
Once `style` is converted to a `RenderStyle`, you will be able to just call `.is_draw`.

gonzalobarbaran replied:
Fixed

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -1266,17 +1267,116 @@ def rect(self, x, y, w, h, style=None):
+        new_color = self.fill_color.colors
+
+        self.set_draw_color(new_color[0] * 255)
+        if style != "D":
```
This means that style is either `FILL` or `DRAW_FILL`.
Once `style` is converted to a `RenderStyle`, you will be able to just call `style.is_fill`.

gonzalobarbaran replied:
Fixed

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -1266,17 +1267,116 @@ def rect(self, x, y, w, h, style=None):
+            point_6 = (x + r, y + h)
+            point_7 = (x, y + h - r)
+
+        original_color = self.draw_color.colors
```
I'm sorry but I really don't understand why you have to alter the _draw color_...
Could you please explain a bit why things do not work with the color currently set?
Also, if ever changing the _draw color_ is needed, it should be reverted back to its original value at the end of this method call.

gonzalobarbaran replied:
The logic behind the color chance is to draw a polygon that fills the rectangle when the style is fill. The problem with that is that there is some weird lines between the arcs. In order to solve that I change the draw color to equal fill color so that weird lines dissapears.

Also, you're right, there was a little mistake in the logic and the first change of color shoud be inside the next if. I fixed that, so it changes the draw color and come back to the original color in the same if

gmischler replied:
> The logic behind the color chance is to draw a polygon that fills the rectangle when the style is fill. The problem with that is that there is some weird lines between the arcs. In order to solve that I change the draw color to equal fill color so that weird lines dissapears.

Just draw the polygon with fill only.
With your recent change to `.is_fill` and `.is_draw`, the horizontal and vertical lines are already drawn seperately as appropriate.

@Lucas-C , Independently of this particular PR.: Would it make things simpler or more complicated to have all the shape methods use the drawing module instead of fiddling with these things locally?

Lucas-C replied:
> Would it make things simpler or more complicated to have all the shape methods use the drawing module instead of fiddling with these things locally?

In terms of "pure" software engineering, yes it would be "cleaner" to use the [drawing](https://pyfpdf.github.io/fpdf2/Drawing.html) instead of directly writing to the internal byte buffer with `FPDF._out`.
But for me this is not a strict requirement as currently most of `FPDF` methods use the 2nd approach. Especially in the case of this PR where the goal is to enhance an exhisting method. A refactor could be done later on, to rewrite some methods to use the `drawing` module, if need be.

Lucas-C commented on fpdf/enums.py:
```diff
@@ -574,5 +574,12 @@ class PDFStyleKeys(Enum):
 
 
+class Corner(CoerciveEnum):
+    TOP_RIGHT = Name("TOP_RIGHT")
```
`Name()` should only be used for string values that are directly embedded in the PDF.
Here you can just use a string ( `"TOP_RIGHT"` )

gonzalobarbaran replied:
Fixed

Lucas-C commented on test/shapes/test_rectangle.py:
```diff
@@ -86,3 +86,118 @@ def test_rect_fill_color(tmp_path):
+    )
+    pdf.rect(160, y, 10, 10, round_corners=("TOP_LEFT", "BOTTOM_LEFT"), style="F")
+
+    assert_pdf_equal(pdf, HERE / "class_round_corners_rect.pdf", tmp_path)
```
Those tests are great!
Could you please add a few mores test cases:
* checking that an exception is raised when an invalid `round_corners` value is provided
* cases for all those conditionals `if r < 0`, `if r == 0`, `if r >= min / 2`
* checking the method behaviour when a different _draw color_ is set

gonzalobarbaran replied:
Yeah, sure

Lucas-C reviewed (changes requested):
Thank you for submitting this PR!
The code quality & tests are great.
I gave you a few feedbacks to improve some details.

PS: Loved the cat picture, thank you ^^

Lucas-C commented:
@allcontributors please add @gonzalobarbaran for code

Lucas-C commented:
Also, a couple of minor things are missing in this PR:
* a mention of this addition in `CHANGELOG.md`
* a short piece of documentation in `docs/Shapes.md`, in the existing "Rectangle" section

Would you mind adding those as part of this PR please?

gmischler commented on fpdf/fpdf.py:
```diff
@@ -1266,17 +1267,116 @@ def rect(self, x, y, w, h, style=None):
+
+    def _draw_rounded_rect(self, x, y, w, h, style, round_corners):
+        min = h
+        r = (w - h) / 2
```
The closer to a square, the smaller the radius? Unless it's exactly a square? That's some rather surprising logic... :wink:

I think you want to add a `corner_radius=#` argument for the user to control the appearance instead. And whatever the default is going to be needs to be something easy to understand and document (a percentage of the smaller side might work, maybe with an absolute upper limit).

gonzalobarbaran replied:
Yeah, I thought about that, but I couldn't figure out how to do it in a proper way.

Lucas-C commented:
Looks like some unit tests are currently failing:
```
 FAILED test/test_opacity.py::test_transparency - ValueError: D is not a valid...
FAILED test/barcodes/test_barcodes.py::test_code39 - ValueError: F is not a v...
FAILED test/barcodes/test_barcodes.py::test_interleaved2of5 - ValueError: F i...
FAILED test/template/test_flextemplate.py::test_flextemplate_multipage - Valu...
FAILED test/template/test_flextemplate.py::test_flextemplate_rotation - Value...
FAILED test/template/test_flextemplate.py::test_flextemplate_elements - Value...
FAILED test/template/test_template.py::test_template_nominal_hardcoded - Valu...
FAILED test/template/test_template.py::test_template_nominal_csv - ValueError...
FAILED test/template/test_template.py::test_template_multipage - ValueError: ...
FAILED test/template/test_template.py::test_template_code39 - ValueError: F i...
FAILED test/template/test_template.py::test_template_code39_legacy - ValueErr...
FAILED test/template/test_template.py::test_template_code39_defaultheight - V...
================= 12 failed, 904 passed, 17 warnings in 51.68s =================
```

gmischler commented:
> Looks like some unit tests are currently failing:
> 
> ...
>  `ValueError: F is not a valid Corner`

Yes, that's the same problem we recently had with new parameters added in the middle of a method signature:

```python
def rect(self, x, y, w, h, round_corners=False, style=None):
```

This breaks all existing code that uses `"F"` as a positional argument instead of  `style="F"`. You'll have to move `round_corners=False` and `corner_radius=#` to the end of the list to prevent that from happening.

Lucas-C commented:
Take your time if you want to finish this @gonzalobarbaran,
but feel free to also ask for help or hand it over as it is, if you need/want to 😊

gonzalobarbaran commented:
sorry @Lucas-C I was very busy this days with work and college, but this weekend i will try to finish this

Lucas-C commented:
Great, no worries, there's no hurry 😊

Lucas-C commented:
Hi @gonzalobarbaran!
How are you doing? 😊
Do you think you may find the motivation & time required to finish this?
You did an great job and I'd be happy to help by doing the rebasing on the `master` branch,
or even handling all the final details if you want!

gonzalobarbaran commented:
Hi, i'm fine. Just give a few days more and i'll finish this

gonzalobarbaran commented:
I think I fixed all

Lucas-C commented:
Great, thank you @gonzalobarbaran!

There was 2 `generate=True` left in the tests 😅
I removed them and fixed the minor merge issue in the `CHANGELOG.md`

Lucas-C commented:
This is merged!

Thank you a lot for your contribution @gonzalobarbaran! 😃

gonzalobarbaran commented:
> Great, thank you @gonzalobarbaran!
> 
> There was 2 `generate=True` left in the tests 😅 I removed them and fixed the minor merge issue in the `CHANGELOG.md`

Oh sorry, I think I forgot to save the file before commiting ir, my bad

gonzalobarbaran commented:
> This is merged!
> 
> Thank you a lot for your contribution @gonzalobarbaran! 😃

Thanks, I hope i can contribute more in the future. Thank you for your kindness and patience :D

Lucas-C commented:
My pleasure 😃

This feature has been included in the new [v2.5.5](https://github.com/PyFPDF/fpdf2/releases/tag/2.5.5) release!
